pyfr/plugins/ascent.py

```diff
@@ -0,0 +1,425 @@
+                configs = self.cfg.getliteral(cfgsect, k)
+                configs = [configs] if isinstance(configs, dict) else configs
+                pn = k.removeprefix('pipeline-')
+
+                pipel_cfg[pn] = configs
+                for j, filter in enumerate(configs):
+                    params = ConduitNode(self.conduit)
+                    for kf, vf in filter.items():
+                        if kf == 'type':
+                            ptype = vf
+                            pl[f'pl_{pn}/f{j}/{kf}'] = vf
+                        elif kf == 'output_name':
+                            if vf in  self._expr:
+                                raise KeyError(f"Output name '{vf}' already used")
+                            params[kf] = vf
+                            self._expr[vf] = None
+                        elif kf == 'field' and vf not in self._expr:
+                            raise KeyError(f"Missing field '{vf}' used in pipeline")
+                        else:
+                            params[kf] = vf
+
+                    pl[f'pl_{pn}/f{j}/params'] = params
+                    if ptype in self.v_filter and 'velocity' not in self._expr:
+                        raise KeyError("Missing implicit 'velocity' field")
+
+        self._pipel_cfg = pipel_cfg
+
+    def _eval_exprs(self, intg):
+        # Get the primitive variable names
+        pnames = self.elementscls.privarmap[self.ndims]
+
+        # Compute the gradients
+        if self._gradpinfo:
+            grad_soln = intg.grad_soln
+
+        # Iterate over each element type in the simulation
+        for idx, ename, rgn in self._ele_regions:
+            # Subset and transpose the solution
+            soln = intg.soln[idx][..., rgn].swapaxes(0, 1)
+
+            d_str = f'domain_{self.dom_0 + idx}'
+
+            _, soln_op = self.ops[ename]
+            # Interpolate the solution to the subdivided points
+            soln = soln_op @ soln
+
+            # Convert from conservative to primitive variables
+            psolns = self.elementscls.con_to_pri(soln, self.cfg)
+
+            # Prepare the substitutions dictionary
+            subs = dict(zip(pnames, psolns), t=intg.tcurr)
+
+            # Prepare any required gradients
+            if self._gradpinfo:
+                grads = np.rollaxis(grad_soln[idx], 2)[..., :]
+
+                # Interpolate the gradients to the points
+                grad_soln = soln_op @ grad_soln
+
+                # Transform from conservative to primitive gradients
+                pgrads = self.elementscls.grad_con_to_pri(soln, grads,
+                                                          self.cfg)
+
+                # Add them to the substitutions dictionary
+                for pname, idx in self._gradpinfo:
+                    for dim, grad in zip('xyz', pgrads[idx]):
+                        subs[f'grad_{pname}_{dim}'] = grad
+
+            for field, expr in self._expr.items():
```

Comment:
In some pipelines you can set an output name for the data produced, to prevent a clash, these get reserved, but don't have an expression connected to them

Reply:
Makes sense.